README
     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================


                	HadrHN test     

                  V. Ivanchenko, A. Bagulya, B. Kutsenko

The goal of this test is to identify accuracy of Geant4 elementary cross sections 
versus PDG collection of data and to control the hydrogen cross section from cross 
sections classes. These elementary cross sections are used to compute cross section 
off nuclei and are also used in hadronic models.

Projectile: proton, neutron, pi+, pi-, kaon+, kaon-, gamma
Elementary cross sections: “NS”, “PDG16”, “SAID”
Cross sections classes: “CHIPS”, “BGG”, “XS”
 
Data are from  http://pdg.lbl.gov.
Files with data are located in HadrHN/data directory. 

	        Compilation
G4BLOCAL points to directory where geant4-dev is located
HADSRC -> .../g4tests-verification/hadronic
G4BWORK points to directory where G4 executables are located (.../bin)
G4BBUILD points to build directory for geant4 (.../geant4-build-mt/)
G4VERSION ->  (10.5.0)

mkdir -p ${G4BLOCAL}/tests-build/HadrHN
cd ${G4BLOCAL}/tests-build/HadrHN
rm -f HadrHN

cmake -DCMAKE_INSTALL_PREFIX=${G4BLOCAL} \
      -DGeant4_DIR=${G4BBUILD}/lib64/Geant4-${G4VERSION} ${HADSRC}/HadrHN
make -j2

cd ${G4BWORK}
rm -f HadrHN
ln -s ${G4BLOCAL}/tests-build/HadrHN/HadrHN

cd ${HADSRC}/HadrHN

                Run Productuion

Simple run for proton, neutron:
$G4BWORK/HadrHN p NS [PDG16/SAID/BGG/CHIPS/XS] H
Simple run for pi+, pi-:
$G4BWORK/HadrHN pi+ NS [PDG16/SAID/BGG/CHIPS] H
Simple run for kaon:
$G4BWORK/HadrHN kaon+ kNS [VG/GG/CHIPS] H
  here VG - work only for K+-
       GG - for all kaons using VG
$G4BWORK/HadrHN alpha XS [ION] H

Simple run for gamma (different targets implemented):
$G4BWORK/HadrHN gamma IAEA [XS/ISOIAEA/ISOXS] H[Symbol of the element]

               Histogramms
ROOT files are creating automatically in HadrHN/test directory.

		Plots
python plotXS.py proton_proton_NS.root 0 # each ROOT file from HadrHN/test directory
python plotXS.py proton_proton_NS.root 1 # "1" only for p+p interaction to add cosmic data


Two models comparison, experimental points added if possible :
python plotXSmodel_only.py gamma_proton_CHIPS_1_0.root gamma_proton_IAEA_1_0.root 0 # Plot from 0 to 200 MeV in linear scale on x and logarithmic scale on y
python plotXSmodel_only.py gamma_proton_CHIPS_1_0.root gamma_proton_IAEA_1_0.root 1 # Plot from 0 to 100 000 MeV in double log scale
python plotXSmodel_only.py gamma_proton_CHIPS_1_0.root gamma_proton_IAEA_1_0.root 2 # Plot from 0 to 1.e+13 MeV in 10-based logarithmic scale on y
 
Plots are creating in HadrHN/plotXS directory.



History
$Id: History,v 1.92 2010-11-09 21:26:24 asaim Exp $
----------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    HadrHN History file
                    ---------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------

06-08-21 B. Kutsenko (hadrhn-V10-06-01)
- added plotXScompare_models.py and plotXSmodel_only.py
- added target choice for the gamma projectile
- README updated
- for Geant4 11b

18-04-19 A.Bagulia (hadrhn-V10-05-02)
- added README

09-04-19 A.Bagulia (hadrhn-V10-05-01)
- updated HadrHN.cc, plotXS.py

29-11-18 V.Ivanchenko (hadrhn-V10-04-01)
- for Geant4 10.5

22-05-18 V.Ivanchenko (hadrhn-V10-04-00)
     


HadrHN.cc

//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
// -------------------------------------------------------------------
//      GEANT 4 class file --- Copyright CERN 1998
//      CERN Geneva Switzerland
//
//
//      File name:     HadrHN.cc
//
//      Author:        V.Ivanchenko
//
//      Creation date: 14 August 2018
//
//      Modifications:
// -------------------------------------------------------------------

#include "globals.hh"
#include "G4ios.hh"
#include <fstream>
#include <iomanip>
#include <algorithm>

#include "G4Material.hh"
#include "G4ElementVector.hh"
#include "G4ProcessManager.hh"
#include "G4VParticleChange.hh"
#include "G4ParticleChange.hh"

#include "G4ParticleTable.hh"
#include "G4IonTable.hh"
#include "G4DynamicParticle.hh"
#include "G4Electron.hh"
#include "G4Positron.hh"
#include "G4Gamma.hh"
#include "G4Proton.hh"
#include "G4Neutron.hh"
#include "G4PionPlus.hh"
#include "G4PionMinus.hh"
#include "G4KaonPlus.hh"
#include "G4KaonMinus.hh"
#include "G4Deuteron.hh"
#include "G4Triton.hh"
#include "G4He3.hh"
#include "G4Alpha.hh"
#include "G4GenericIon.hh"

#include "G4ForceCondition.hh"
#include "G4Box.hh"
#include "G4PVPlacement.hh"
#include "G4Step.hh"
#include "G4GRSVolume.hh"

#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"

#include "G4StateManager.hh"
#include "G4NistManager.hh"
#include "G4Material.hh"

#include "Histo.hh"
#include "G4Timer.hh"

#include "G4HadronNucleonXsc.hh"
#include "G4ComponentSAIDTotalXS.hh"
#include "G4ParticleInelasticXS.hh"
#include "G4NeutronInelasticXS.hh"
#include "G4NeutronElasticXS.hh"
#include "G4ChipsNeutronElasticXS.hh"
#include "G4ChipsNeutronInelasticXS.hh"
#include "G4ChipsProtonElasticXS.hh"
#include "G4ChipsProtonInelasticXS.hh"
#include "G4ChipsPionPlusElasticXS.hh"
#include "G4ChipsPionPlusInelasticXS.hh"
#include "G4ChipsPionMinusElasticXS.hh"
#include "G4ChipsPionMinusInelasticXS.hh"
#include "G4ChipsKaonPlusElasticXS.hh"
#include "G4ChipsKaonPlusInelasticXS.hh"
#include "G4ChipsKaonMinusElasticXS.hh"
#include "G4ChipsKaonMinusInelasticXS.hh"
#include "G4BGGNucleonElasticXS.hh"
#include "G4BGGNucleonInelasticXS.hh"
#include "G4BGGPionElasticXS.hh"
#include "G4BGGPionInelasticXS.hh"
#include "G4PhotoNuclearCrossSection.hh"
#include "G4VCrossSectionDataSet.hh"
#include "G4eCoulombScatteringModel.hh"
#include "G4EmParameters.hh"
#include "G4ComponentGGNuclNuclXsc.hh"
#include "G4GammaNuclearXS.hh"
#include "G4Isotope.hh"
#include "G4Element.hh"
#include "G4Material.hh"

#include "G4NistManager.hh"
#include "G4SystemOfUnits.hh"
#include "G4PhysicalConstants.hh"

int main(int argc, char** argv)
{
  // Control on input
  if(argc < 4) {
    G4cout << "Input parameters are not specified! Exit" << G4endl;
    return 1;
  }

  G4cout << "=======================================================" << G4endl;
  G4cout << "======   Hadron Nucleon Cross Section Test     ========" << G4endl;
  G4cout << "=======================================================" << G4endl;

  // ------- Initialisation 

  G4int     verbose  = 1, ncomponents, natoms, Z , A=0;
  G4double  energy   = 1.*MeV, density;
  G4String name = argv[2],atomName = argv[3];
  G4MaterialCutsCouple* isocouple=new G4MaterialCutsCouple();
  G4ProductionCuts* pcut = new G4ProductionCuts();
  pcut->SetProductionCut(0.0, 1);
  pcut->SetProductionCut(0.0, 2);
  pcut->SetProductionCut(0.0, 3);
  pcut->SetProductionCut(DBL_MAX, 4);

  const G4Material* mat = G4NistManager::Instance()->FindOrBuildMaterial("G4_"+atomName);
  G4MaterialCutsCouple* couple = new G4MaterialCutsCouple(mat, pcut);
  Z = mat->GetZ();
  
  if(name == "ISOXS"||name == "ISOCHIPS" ){
  G4cout<<"Enter isotop A: "<<G4endl;
  G4cin>> A;
  if(Z<3||(Z>4&&Z<11)||(Z>13&&Z<19)||(Z>31&&Z<37)||(Z>50&&Z<55)||(Z>83&&Z<87)) natoms = 2;
  else natoms =1;
  G4Isotope* iso = new G4Isotope(atomName, Z, A);
  G4Element* elIso = new G4Element("Isotop_"+atomName ,atomName, ncomponents=1);
  elIso->AddIsotope(iso, 1.0);
  G4double    pressure = atmosphere, temperature = 293*kelvin, molar_constant = CLHEP::Avogadro*CLHEP::k_Boltzmann;
  density = (elIso->GetA()*pressure)/(temperature*molar_constant);
  G4Material* mIso = new G4Material(atomName+"_"+std::to_string(natoms), density, ncomponents=1);
  mIso->AddElement(elIso, natoms);
  isocouple->SetMaterial(mIso);
  isocouple->SetProductionCuts(pcut);
  isocouple->SetIndex(1); 
  }
  else 
  couple->SetIndex(0);
 
  const G4int nidx = 11;
  const G4String particle[nidx] = {"p", "n", "pi+", "pi-", "kaon+", "kaon-", "gamma",
                                   "deuteron","triton","He3","alpha"};
  const G4ParticleDefinition* pDef[nidx] = {
    G4Proton::Proton(),G4Neutron::Neutron(),G4PionPlus::PionPlus(),G4PionMinus::PionMinus(),
    G4KaonPlus::KaonPlus(),G4KaonMinus::KaonMinus(),G4Gamma::Gamma(),
    G4Deuteron::Deuteron(),G4Triton::Triton(),G4He3::He3(),G4Alpha::Alpha() };
  std::vector<G4String> p_mod_names = {"NS", "PDG16", "SAID", "CHIPS", "BGG", "XS"};
  std::vector<G4String> n_mod_names = {"NS", "PDG16", "SAID", "CHIPS", "BGG", "XS"};
  std::vector<G4String> pip_mod_names = {"NS", "PDG16", "SAID", "CHIPS", "BGG"};
  std::vector<G4String> pin_mod_names = {"NS", "PDG16", "SAID", "CHIPS", "BGG"};
  std::vector<G4String> kp_mod_names = {"kNS", "VG", "GG", "CHIPS"};
  std::vector<G4String> kn_mod_names = {"kNS", "VG", "GG", "CHIPS"};
  std::vector<G4String> gp_mod_names = {"CHIPS","XS","ISOCHIPS","ISOXS"};
  std::vector<G4String> ip_mod_names = {"NN","XS","ION"};
  std::vector<G4String> part_mod_names;

  const G4ParticleDefinition* project = nullptr;  
  G4String partname = argv[1];
  G4int idx = 0;
  for (; idx < nidx; idx++) {
    if (partname == particle[idx]) {
      project = pDef[idx];
      if(0 == idx) { part_mod_names = p_mod_names; }
      else if(1 == idx) { part_mod_names = n_mod_names; }
      else if(2 == idx) { part_mod_names = pip_mod_names; }
      else if(3 == idx) { part_mod_names = pin_mod_names; }
      else if(4 == idx) { part_mod_names = kp_mod_names; }
      else if(5 == idx) { part_mod_names = kn_mod_names; }
      else if(6 == idx) { part_mod_names = gp_mod_names; }
      else { part_mod_names = ip_mod_names; }
      break;
    }
  }
  if(!project) {
    G4cout << "Projectile <" << partname << "> it's not implemented yet! Exit" << G4endl;
    return 1;
  }
  
  
  G4VCrossSectionDataSet* inel = nullptr;
  G4VCrossSectionDataSet* el = nullptr;
  const G4ParticleDefinition* target;
  if(part_mod_names != gp_mod_names &&  strcmp(atomName,"H")!=0)  {
    G4cout << "Collisions with different materials implemented only for gamma. For other particles use H!\nExit" << G4endl;
    return 0;
  }
  else target = pDef[0];


  G4HadronNucleonXsc hnxs;
  G4ComponentGGNuclNuclXsc* nnxs = new G4ComponentGGNuclNuclXsc();
  G4ComponentSAIDTotalXS* said = new G4ComponentSAIDTotalXS();
  G4eCoulombScatteringModel* ss = new G4eCoulombScatteringModel(false);


  
  if(std::find(part_mod_names.begin(),part_mod_names.end(),name)==part_mod_names.end()){
    G4cout << "For projectile <" << partname <<"" "> Model <"<<name<<"> is not implemented. Exit" << G4endl;
    return 1;
  }

  G4int  isHNX = -1;
  
  if (name == "CHIPS" || name == "ISOCHIPS") {
    if (0 == idx) {
      el = new G4ChipsProtonElasticXS();
      inel = new G4ChipsProtonInelasticXS();
    }
    else if (1 == idx) {
      el = new G4ChipsNeutronElasticXS();
      inel = new G4ChipsNeutronInelasticXS();
    }
    if (2 == idx) {
      el = new G4ChipsPionPlusElasticXS();
      inel = new G4ChipsPionPlusInelasticXS();
    }
    else if (3 == idx) {
      el = new G4ChipsPionMinusElasticXS();
      inel = new G4ChipsPionMinusInelasticXS();
    }
    if (4 == idx) {
      el = new G4ChipsKaonPlusElasticXS();
      inel = new G4ChipsKaonPlusInelasticXS();
    }
    else if (5 == idx) {
      el = new G4ChipsKaonMinusElasticXS();
      inel = new G4ChipsKaonMinusInelasticXS();
    }
    else if (6 == idx) {
      inel = new G4PhotoNuclearCrossSection();
    }
  } else if (name == "BGG") {
    if (0 == idx || 1 == idx) {
      el = new G4BGGNucleonElasticXS(project);
      inel = new G4BGGNucleonInelasticXS(project);
    } else if(2 == idx || 3 == idx) {
      el = new G4BGGPionElasticXS(project);
      inel = new G4BGGPionInelasticXS(project);
    }
  }else if (name == "XS"|| name == "ISOXS") {
    if (0 == idx) {
      inel = new G4ParticleInelasticXS(G4Proton::Proton());
      el = new G4BGGNucleonElasticXS(project);
    }
    else if (1 == idx) {
      el = new G4NeutronElasticXS();
      inel = new G4NeutronInelasticXS();
    }
    else if (6 == idx) {
      inel = new G4GammaNuclearXS();
    }
    else if (7 == idx) {
      inel = new G4ParticleInelasticXS(G4Deuteron::Deuteron());
    }
    else if (8 == idx) {
      inel = new G4ParticleInelasticXS(G4Triton::Triton());
    }
    else if (9 == idx) {
      inel = new G4ParticleInelasticXS(G4He3::He3());
    }
    else if (10 == idx) {
      inel = new G4ParticleInelasticXS(G4Alpha::Alpha());
    }
  }
  else if(name == "NS") { 
    isHNX = 0; 
  } else if(name == "PDG16") { 
    isHNX = 2;
  } else if(name == "SAID") { 
    isHNX = 3;
  } else if(name == "kNS") { 
    isHNX = 4;
  } else if(name == "VG") { 
    isHNX = 5;
  } else if(name == "GG") { 
    isHNX = 6;
  } else if(name == "NN") { 
    isHNX = 7;
  } 

  if(isHNX < 0 && !inel) {
    G4cout << "Cross-section model <" << name 
	   << "> does not exist! Exit" << G4endl;
    return 1;
  }

  if (inel) inel->BuildPhysicsTable(*project);
  if (el) el->BuildPhysicsTable(*project);
  G4DataVector dv;
  dv.resize(4, DBL_MAX);
  G4EmParameters::Instance()->SetMscThetaLimit(0.0);
  ss->SetPolarAngleLimit(0.0);
  ss->Initialise(project, dv);
  if(name == "ISOXS" || name == "ISOCHIPS") ss->SetCurrentCouple(isocouple);
  else  ss->SetCurrentCouple(couple);
  
  
  G4ThreeVector aPosition = G4ThreeVector(0.,0.,0.);
  G4ThreeVector aDirection = G4ThreeVector(0.0,0.0,1.0);

  G4DynamicParticle dParticle(project,aDirection,energy);

  G4cout.setf( std::ios::scientific, std::ios::floatfield );
  G4int  prec = G4cout.precision(3);

  G4ParticleTable* partTable = G4ParticleTable::GetParticleTable();
  partTable->SetReadiness();

  // ------- Histograms name 
  Histo    histo;
  G4String hname;
  if(strcmp(atomName,"H")==0 && A != 2) hname = "test/" + project->GetParticleName() + "_" 
    + target->GetParticleName() + "_" + name + "_" +std::to_string(Z)+"_" +std::to_string(A) ;
  else if (strcmp(atomName,"H")==0 && A == 2)hname = "test/" + project->GetParticleName() + "_" 
    +  "deuteron_" + name + "_" +std::to_string(Z)+"_" +std::to_string(A) ;
  else hname = "test/" + project->GetParticleName() + "_" 
	 + atomName + "_" + name + "_" +std::to_string(Z)+"_" +std::to_string(A) ;
  G4double mass = project->GetPDGMass()/MeV;

  const G4int nbins = 48001;
  G4double pmin =  10*MeV;
  G4double pmax =  1e10*GeV;
  G4double xmin =  std::log10(pmin);
  G4double xmax =  std::log10(pmax);
  G4double dx = (xmax - xmin)/G4double(nbins-1);
  
  histo.Add1D("0","Inelastic",nbins,xmin-dx*0.5,xmax+dx*0.5);
  histo.Add1D("1","Elastic",nbins,xmin-dx*0.5,xmax+dx*0.5);
  histo.Add1D("2","Total",nbins,xmin-dx*0.5,xmax+dx*0.5);
  histo.Add1D("3","EM Elastic",nbins,xmin-dx*0.5,xmax+dx*0.5);

  G4double lxmin = 0*MeV;
  G4double lxmax = 1e5*MeV;
  G4double dlx = (lxmax - lxmin)/G4double(nbins-1);
  
  if(partname == "gamma") histo.Add1D("4","Photonuclear Low Energy Region",nbins,lxmin-dlx*0.5,lxmax+dlx*0.5);
  histo.SetFileName(hname);
  histo.Book();
  G4cout << "Histograms are booked output file <" << hname << "> "
	 << G4endl;

  G4double xse[nbins],xsi[nbins],xst[nbins],em[nbins],lxsi[nbins];

  G4int i;
  dParticle.SetDefinition(project);
  G4double x0 = xmin;
  G4double lx0 = lxmin;
  if(verbose>1) { 
    G4cout << "----  "
	   << project->GetParticleName() << " + " << atomName
	   << " -------------" << G4endl;
    G4cout << "  PLAB        EKIN       XSIN        XSEL      XSTOT    EM    log(p/770)  log10(p)\n"
	   << "------------------------------------------------------------------------------------"
	   << G4endl; 
  }

  for (i=0; i<nbins; ++i) {
    G4double p = std::pow(10., x0)*MeV;
    G4double lp = lx0;
    lx0 += dlx;
    x0 += dx;
    G4double e = std::sqrt(p*p+mass*mass)-mass;
    G4double le = std::sqrt(lp*lp+mass*mass)-mass;
    //if(6 == idx && e > 20.*GeV && name == "XS") { e = 20.*GeV; }
    dParticle.SetKineticEnergy(e);
    xse[i] = 0.;
    if(isHNX < 0) {
      if (idx == 6 && (name == "CHIPS" || name == "XS")){
	xsi[i] = inel->GetElementCrossSection(&dParticle,mat->GetZ(),0)/millibarn;
	dParticle.SetKineticEnergy(le);
	lxsi[i] = inel->GetElementCrossSection(&dParticle,mat->GetZ(),0)/millibarn;
	dParticle.SetKineticEnergy(e);
      } else if (idx == 6 && (name == "ISOCHIPS" || name == "ISOXS")){
	xsi[i] = inel->GetIsoCrossSection(&dParticle,Z,A)/millibarn;
	dParticle.SetKineticEnergy(le);
	lxsi[i] = inel->GetIsoCrossSection(&dParticle,Z,A)/millibarn;
	dParticle.SetKineticEnergy(e);
      } else if (idx > 6) {
	xsi[i] = inel->GetElementCrossSection(&dParticle,1,0)/millibarn;
      } else if (name == "CHIPS") {
	xsi[i] = inel->GetIsoCrossSection(&dParticle,1,1)/millibarn;
	xse[i] = el->GetIsoCrossSection(&dParticle,1,1)/millibarn;
      } else {
	xsi[i] = inel->GetElementCrossSection(&dParticle,1,0)/millibarn;
	xse[i] = el->GetElementCrossSection(&dParticle,1,0)/millibarn;
      }
    }
    else if (isHNX == 0) {
      hnxs.HadronNucleonXscNS(project,target,e);
      xsi[i] = hnxs.GetInelasticHadronNucleonXsc()/millibarn;
      xse[i] = hnxs.GetElasticHadronNucleonXsc()/millibarn;
    }
    else if (isHNX == 2) {
      hnxs.HadronNucleonXscPDG(project,target,e);
      xsi[i] = hnxs.GetInelasticHadronNucleonXsc()/millibarn;
      xse[i] = hnxs.GetElasticHadronNucleonXsc()/millibarn;
    }
    else if (isHNX == 3) {
      hnxs.HadronNucleonXscPDG(project,target,e);
      xsi[i] = said->GetInelasticIsotopeCrossSection(project,e,1,1)/millibarn;
      xse[i] = said->GetElasticIsotopeCrossSection(project,e,1,1)/millibarn;
    }
    else if (isHNX == 4) {
      hnxs.KaonNucleonXscNS(project,target,e);
      xsi[i] = hnxs.GetInelasticHadronNucleonXsc()/millibarn;
      xse[i] = hnxs.GetElasticHadronNucleonXsc()/millibarn; 
    }
    else if (isHNX == 5) {
      hnxs.KaonNucleonXscVG(project,target,e);
      xsi[i] = hnxs.GetInelasticHadronNucleonXsc()/millibarn;
      xse[i] = hnxs.GetElasticHadronNucleonXsc()/millibarn;
    }
    else if (isHNX == 6) {
      hnxs.KaonNucleonXscGG(project,target,e);
      xsi[i] = hnxs.GetInelasticHadronNucleonXsc()/millibarn;
      xse[i] = hnxs.GetElasticHadronNucleonXsc()/millibarn;
    }
    else if (isHNX == 7) {
      xsi[i] = nnxs->GetInelasticElementCrossSection(project,e,1,1)/millibarn;
      xse[i] = nnxs->GetElasticElementCrossSection(project,e,1,1)/millibarn;
    }
    xst[i] = xsi[i] + xse[i];

    em[i] = 
      ss->ComputeCrossSectionPerAtom(project, e, 1.0, 1.0, 0.0, e)/millibarn;
    
    if(verbose>1) { 
      G4cout << i << ". " << p << "  ekin= " << e << "   " << xsi[i] 
             << "  " << xse[i] << "  " << xst[i] << "  " << em[i]
	     << "  " << G4Log(p/770.) 
	     << "  " << std::log10(p) << G4endl;
    }

    histo.Fill(0,x0,xsi[i]);
    histo.Fill(1,x0,xse[i]);
    histo.Fill(2,x0,xst[i]);
    //histo.Fill(3,x0,em[i]);
    if(idx == 6) histo.Fill(4,lx0,lxsi[i]);
  }
  G4cout << "------------------------------------------------------------------------------------"
	 << G4endl;
  G4cout.precision(prec);
  
  if(verbose > 0) { G4cout << "###### Save histograms" << G4endl; }
  histo.Save();
  
  if(verbose > 0) {
    G4cout << "###### End of run # " << G4endl;
  }
  G4double ma = G4PionPlus::PionPlus()->GetPDGMass();
  G4cout << std::sqrt(ma*ma  + 10000) - ma << G4endl;
}


plotXScompare_models.py
#!/usr/bin/env python
#
# Plot XS pictures 
# A. Bagulya
# 23.03.2019
#
# Updated for phototonuclear
# Kutsenko Bogdan
# 20.07.2021




import os, sys
from ROOT import TROOT, gROOT, gStyle, gPad, TCanvas, TStyle, TPad
from ROOT import TFile, TH1, TLegend, TGraphAsymmErrors
from array import array
from math import log10
import errno

def plotXS():
    gROOT.Reset()
    gROOT.SetStyle('Plain')
    gROOT.SetBatch(True)

    if len(sys.argv) < 3:
        print('''Usage: python plotXS.py file.root 0/1/2''')
        sys.exit()
    else:
        fname = sys.argv[1]
        fname2 = sys.argv[2]
        

        flag  = int(sys.argv[3])

    project = os.path.splitext(os.path.basename(fname))[0].split('_')[0]
    target = os.path.splitext(os.path.basename(fname))[0].split('_')[1]
    model_name = os.path.splitext(os.path.basename(fname))[0].split('_')[2]
    Z = os.path.splitext(os.path.basename(fname))[0].split('_')[3]
    A = os.path.splitext(os.path.basename(fname))[0].split('_')[4]

    project2 = os.path.splitext(os.path.basename(fname2))[0].split('_')[0]
    target2 = os.path.splitext(os.path.basename(fname2))[0].split('_')[1]
    model_name2 = os.path.splitext(os.path.basename(fname2))[0].split('_')[2]
    Z2 = os.path.splitext(os.path.basename(fname2))[0].split('_')[3]

    if model_name == 'kNS':
        model_name = 'NS'

    #HADSRC = readEnvVar('HADSRC')
    testpath = os.path.join('./','')
    dirName = os.path.join('plotXS/',str(target))

    
    try:
    # Create target Directory
        os.mkdir(dirName)
        print("Directory " , dirName ,  " Created ") 
    except OSError as e:
        if e.errno == errno.EEXIST:
            print("Directory " , dirName ,  " already exists")
        else:
            raise

        
    plotpath = os.path.join('./','plotXS/'+str(target))
    datapath = os.path.join('./','data')

    
    if (int(flag) == 0 or int(flag == 1)) and project != 'gamma':
        print('Flag 0/1 (Inelastic cross-section from 0 to 200 MeV) implemented only for gamma')
        exit()
        
    part = {'proton':'p', 'neutron':'n', 'pi+':'#pi^{+}', 'pi-':'#pi^{-}', 'kaon+':'K^{+}', 'kaon-':'K^{-}', 'gamma':'#g' , 'deuteron':'d', 'triton':'t', 'He3':'he3', 'alpha':'alpha'}

    global PLAB, SIG, errPLABl, errPLABh, errSIGl, errSIGh

    PLAB, SIG = array ( 'd' , [] ), array ( 'd' , [] )
    errPLABl, errPLABh, errSIGl, errSIGh = array ( 'd' , [] ), array ( 'd' , [] ), array ( 'd' , [] ), array ( 'd' , [] )

    c1 = TCanvas('c1', 'c1',200,100,900,600)
    gStyle.SetOptStat(0)
    c1.SetFillColor(0)
    c1.SetBorderMode(0)
    c1.SetBorderSize(0)
    c1.SetFrameBorderMode(0)

    hasDataEl = 0;
    hasDataIn = 0;
    if project == 'gamma':
        hasDataEl = 0;
    
    if project == 'alpha':
        hasDataEl = 0
        hasDataIn = 0;

    Name = 'Total, elastic and inelastic cross sections ' + project + ' + ' + target + ' in mb, model = ' + model_name
    if project == 'gamma':
        Name = 'Total cross sections gamma' + ' + '+target + A+ ' in mb, model = ' + model_name + ' + ' + model_name2
         
    crossampl = [0,   4,   6,   10,  6,  20,  24,  26,  29,  30,  30,  30,  45,  90,  50,  60,  70,  80,  65,  90, 100,  100,  110,  110,  120,  120,  130,  130,  140,  140, 150,  150,  160,   170,   170,  200,   210,   220,  230,  240,   250, 260,   270,  280,  290, 300, 310, 320, 330, 340, 350,   360, 370,   380, 390, 400,   410,   420,   430,   440,  450,   460,   470, 480,   490,   500,   510,   520,   530, 540,  550,   560, 600, 600,   610,   620,   630, 650, 660,   680, 700, 720,  740,   760,   780,   800,   820,   840,   860, 880, 900]
    lbound0=1.
    rbound0=100000.
    lbound1=1.
    rbound1=200.
    if project == 'neutron':
        hhh = gPad.DrawFrame(1, 1., 7, 10000. ,'cross sections in mb')
    elif project == 'kaon+':
        hhh = gPad.DrawFrame(1, 1., 7, 100. ,'cross sections in mb')
    elif project == 'gamma' and int(flag) == 2:
        hhh = gPad.DrawFrame(0.1, 0.01, 13., crossampl[int(Z)] ,'cross sections in mb')
    elif project == 'gamma' and int(flag) == 1 :
        hhh = gPad.DrawFrame(lbound0, 0.01, rbound0, 4*crossampl[int(Z)] ,'cross sections in mb')
    elif project == 'gamma' and int(flag) == 0:
        hhh = gPad.DrawFrame(lbound1, 0., rbound1, crossampl[int(Z)] ,'cross sections in mb')
    elif project == 'proton':
        hhh = gPad.DrawFrame(1, 1., 12, 1000. ,'cross sections in mb')
    else:
        hhh = gPad.DrawFrame(1, 1., 7, 1000. ,'cross sections in mb')

    hhh.SetTitle(Name)
    if int(flag) == 2:
        hhh.GetXaxis().SetTitle('log10(P/(MeV/c))')
    else:
        hhh.GetXaxis().SetTitle('P, MeV/c')
    
    hhh.GetYaxis().SetTitle('#sigma, mb')
    hhh.GetXaxis().SetTitleOffset(1.2)
    hhh.GetYaxis().SetTitleOffset(1.2)
    hhh.Draw('AXIS SAME')
    if int(flag) == 2:
        gPad.SetLogy(1)
    elif int(flag) == 1 :
        gPad.SetLogx(1)
        gPad.SetLogy(1)
    gPad.SetGrid(1, 1)
    gPad.Update()

    leg = TLegend(.45, .75, 0.97, .9)
    leg.SetTextSize(.03)
    leg.SetHeader('Cross sections') 

    rootflname = os.path.join(testpath, fname)
    print('File to open ', rootflname)
    f1 = TFile(rootflname)
    if int(flag) == 0 or int(flag) == 1:
        AddH1(f1, leg, 2, 2, 'h4', 'Photonuclear Low Energy region ' + model_name)
    else:
        AddH1(f1, leg, 2, 2, 'h0', 'Inelastic ' + model_name)
    c1.Update()
    if project != 'gamma':
        AddH1(f1, leg, 3, 2, 'h1', 'Elastic ' + model_name)
        c1.Update()
        AddH1(f1, leg, 4, 2, 'h2', 'Total ' + model_name)
        c1.Update()
    rootflname = os.path.join(testpath, fname2)
    print('File to open ', rootflname)
    f2 = TFile(rootflname)
    if int(flag) == 0 or int(flag) == 1:
        AddH1(f2, leg, 4, 2, 'h4', 'Photonuclear Low Energy region ' + model_name2)
    else:
        AddH1(f2, leg, 4, 2, 'h0', 'Inelastic ' + model_name2 )


    datafile = os.path.join(datapath, 'rpp2016-' + str(project) + str(target) +'_'+str(Z)+'_'+str(A)+ '_total.dat')
    if os.path.isfile(datafile) and flag == 2:
        AddGraph(datafile, leg, 'data Tot PDG16', 2)
        
    c1.Update()
    leg.Draw()
    if int(flag) == 2:
       fout = os.path.join(plotpath, 'A_' + model_name + '_' + model_name2 +'_'+str(Z)+'_'+str(A)+ '.png')
    elif int(flag) == 0:
       fout = os.path.join(plotpath, 'A_' + model_name + '_' + model_name2 + '_' +str(lbound1)+'_'+str(rbound1) + 'MeV_'+str(Z)+'_'+str(A)+'.png')
    elif int(flag) == 1:
        fout = os.path.join(plotpath, 'A_' + model_name + '_' + model_name2 +'_'+str(lbound0)+'_'+str(rbound0) +'MeV_'+str(Z)+'_'+str(A) +'.png')
    c1.Print(fout)
    c1.Close()

def readEnvVar(envar):
    if os.environ.get(envar) is None:
        print('''No environment variable %s''' % envar)
        sys.exit()
    else:
        return os.environ.get(envar)

def openFile(fileName):
    global infile
    try:
        infile = open(fileName, 'r')
    except IOError:
        print('Input file <',fileName, '> does not exist! Exit')
        sys.exit(2)

def AddH1(ff, leg, idx, wdth, hh, title):
    h1 = ff.Get(hh)
    h1.SetLineColor(idx)
    h1.SetLineWidth(wdth)
    h1.Draw('HISTO SAME C')
    leg.AddEntry(h1, title, 'l')

def readFile(ff):
    openFile(ff)
    lines = infile.readlines()
    num = int(lines[10])
    for line in lines[11:11+num]:
        plab = 1000.*float(line.split()[1])
        pmin = 1000.*float(line.split()[2])
        pmax = 1000.*float(line.split()[3])
        sig = float(line.split()[4])
        errsigh = float(line.split()[5])
        errsigl = float(line.split()[6])

        PLAB.append(log10(plab))
        errPLABl.append(0.)
        errPLABh.append(0.)
        SIG.append(sig)
        errSIGl.append(errsigh)
        errSIGh.append(errsigl)
    infile.close()

def AddGraph(ff, leg, title, idx):
     openFile(ff)
     readFile(ff)
     globals()['gr' + str(idx)] = TGraphAsymmErrors(len(PLAB),PLAB,SIG,errPLABl,errPLABh,errSIGl,errSIGh)
     globals()['gr' + str(idx)].SetMarkerStyle(19+idx)
     globals()['gr' + str(idx)].SetMarkerColor(1)
     globals()['gr' + str(idx)].SetMarkerSize(0.5)
     globals()['gr' + str(idx)].Draw('SAME P')
     leg.AddEntry(globals()['gr' + str(idx)],title,'ep')

     del PLAB[:]
     del errPLABl[:]
     del errPLABh[:]
     del SIG[:]
     del errSIGl[:]
     del errSIGh[:]

###______________________________
if __name__ == '__main__':
    plotXS()
